core/amber/src/main/python/core/architecture/rpc/async_rpc_server.py

```diff
@@ -1,122 +1,124 @@
+        :return: A Stream object that provides asynchronous send and receive methods.
+
+        This allows the incoming ControlRequest to be treated as a streaming request
+        for compatibility with grpclib's handler interface.
+        """
+
+        class ControlRequestStream(grpclib.server.Stream):
+            def __init__(self):
+                self.result = None
+
+            async def recv_message(self):
+                return request
+
+            async def send_message(self, msg):
+                self.result = msg
+
+        return ControlRequestStream()
 
     def receive(
-        self, from_: ActorVirtualIdentity, control_invocation: ControlInvocationV2
+        self, from_: ActorVirtualIdentity, control_invocation: ControlInvocation
     ):
-        command: ControlCommandV2 = get_one_of(control_invocation.command)
+        """
+        Handles incoming ControlInvocation messages by invoking the appropriate handler.
+
+        :param from_: The sender's ActorVirtualIdentity.
+        :param control_invocation: The incoming ControlInvocation message.
+
+        This method performs the following steps:
+        1. Extracts the command from the ControlInvocation.
+        2. Looks up the corresponding handler for the method name.
+        3. Wraps the command as a stream and runs the handler asynchronously.
+        4. Constructs a ControlReturn or ControlError based on the handler's result.
+        5. Sends the response back to the sender, unless no reply is needed.
+        """
+        command: ControlRequest = get_one_of(control_invocation.command)
+        method_name = control_invocation.method_name
         logger.debug(f"PYTHON receives a ControlInvocation: {control_invocation}")
         try:
-            handler = self.look_up(command)
-            control_return: ControlReturnV2 = set_one_of(
-                ControlReturnV2, handler(self._context, command)
+            # Look up the handler based on the lowercase method name.
+            handler: grpclib.const.Handler = self.look_up(method_name.lower())
+            # Wrap the command as a streaming request.
+            control_payload_stream = self.wrap_as_stream(command)
+            # Run the handler asynchronously.
+            asyncio.run(handler.func(control_payload_stream))
+            # Set up a ControlReturn from the handler's result.
+            control_return: ControlReturn = set_one_of(
+                ControlReturn, control_payload_stream.result
             )
 
         except Exception as exception:
+            # Handle exceptions and log the error.
             logger.exception(exception)
-            control_return: ControlReturnV2 = set_one_of(
-                ControlReturnV2, ControlException(str(exception))
+            # Construct a ControlError message in case of an exception.
+            control_return: ControlReturn = set_one_of(
+                ControlReturn,
+                ControlError(
```

Comment:
can we keep the name `ControlException`?

Reply:
I suggest we do it in another PR.